include/jemalloc/internal/sized_alloc_region.h

```diff
@@ -0,0 +1,144 @@
+#ifndef JEMALLOC_INTERNAL_SIZED_ALLOC_REGION_H
+#define JEMALLOC_INTERNAL_SIZED_ALLOC_REGION_H
+
+#include "jemalloc/internal/alloc_ctx.h"
+#include "jemalloc/internal/assert.h"
+#include "jemalloc/internal/atomic.h"
+#include "jemalloc/internal/pages.h"
+
+/*
+ * This file defines sized alloc regions.  These regions have portions of memory
+ * dedicated to a particular size class.  When allocations come out of these
+ * regions, we can quickly determine their size based only on their address and
+ * a little bit of global metadata, thus avoiding any rtree lookups or (if we're
+ * lucky) any other cache-unfriendly lookups at all.
+ *
+ * We orchestrate it so that zero-initialization is a valid state for the
+ * sized_alloc_region_t.  This lets us avoid branching on
+ * opt_sized_alloc_region.
+ */
+
+/*
+ * SIZED_ALLOC_REGION_LG_NUM_SCS:
+ *     The log of the number of size classes that can come out of the
+ *     sized-alloc region.
+ * SIZED_ALLOC_REGION_LG_SC_SIZE:
+ *     The log of the size of virtual address space that we allocate to each
+ *     size class within the sized-alloc region.
+ */
+#if LG_SIZEOF_PTR == 2
```

Comment:
I'm not sure I really understand how decaying works. (Not sure this review is the best place to discuss this either, so if you have a better thought, let me know).

I ended up with a moderately complicated implementation of our hooks, but basically trying to solve the problem that we want to use 2MB TLB entries when we can, and we want to limit syscalls for performance. chnk_alloc pulls from a recycling list, if there's memory present (we cache 2MB and 4MB chunks, and our system rarely allocates larger than this). If there's nothing in the recycler, we mmap in 64MB blocks and push the unused part into the recycler. dalloc just pushes the chunk into our recycling lists.

commit just returns false; decommit just returns true; purge just returns true; split just returns false; merge just returns false. As best I could tell from the 4.2 documentation this is correct.

Our alloc is more complex than I described, because we also want to release some of our recycler's memory when we get a alloc() request over 4MB, if we're near the process's virtual memory limit. There's also some fiddly bits so we can track how much memory we've mmap'd and whether it was HUGETLB or 4k mappings, so we can report on this occasionally in the logs.

We had to tweak Linux to disable transparent hugepages, because it caused a significant performance issue where our mmap() calls would hang for a while; we can't afford to have threads hanging for 100ms or more in a syscall. So we manually take about 2/3rds of the DRAM on the box and pre-reserve hugepages, and then our chunk hooks will take those if available and fall back to normal 4k mappings. My investigation into MADV_HUGETLB showed I had no way to tell if I really got a 2MB mapping or not, and I really want to be able to report in the application how much data is backed by 2MB TLBs. I didn't spend a ton of time testing the performance of this versus the pre-reserved MAP_HUGETLB.

I agree MAP_HUGETLB is really hard -- how would jemalloc know if the pre-allocated large pages were meant for jemalloc or for something else in the system?